### PR TITLE
Spells refactoring7

### PR DESCRIPTION
**Pull request by alexvins**
Re-factoring never ends:)

Subject of this PR are healing/rising spells

### CODE REVIEW

**Comment by alexvins** [comment]
Enough refactoring for now. Only bug fixes will go to this branch.

**Comment by alexvins** [comment]
[COMPLETE FREZE] No more mantiss issues will go here. Only regression fixes. Ready to review & testing.

**Comment by ArseniyShestakov** [comment]
Found one regression:
http://bugs.vcmi.eu/view.php?id=2289

Posted it on bug tracker initially as had no idea branch may cause anything like that.

**Comment by ArseniyShestakov** [comment]
2nd regression: archangels now can't cast resurrection on other archangels stack.

**Comment by alexvins** [comment]
First one has been fixed.

File: lib/NetPacksLib.cpp [hunk 1]
```diff
@@ -1459,21 +1459,15 @@ DLL_LINKAGE void StacksHealedOrResurrected::applyGs( CGameState *gs )
...
+			//removing all effects from negative spells
+			auto selector = [](const Bonus * b)
 			{
 				const CSpell *s = b->sourceSpell();
-				if(s && s->isNegative())
-				{
-					changedStack->removeBonus(b);
-				}
-			}
+				return (s != nullptr) && s->isNegative() && (s->id != SpellID::DISRUPTING_RAY);
```
Comment (DjWarmonger) [inline lib/NetPacksLib.cpp]: Is't it the case when "PERMANENT" effect duration should be used?

File: lib/spells/CDefaultSpellMechanics.cpp [hunk 2]
```diff
@@ -849,24 +736,125 @@ std::set<const CStack *> DefaultSpellMechanics::getAffectedStacks(SpellTargeting
...
 void DefaultSpellMechanics::doDispell(BattleInfo * battle, const BattleSpellCast * packet, const CSelector & selector) const
 {
+	auto localSelector = [](const Bonus * bonus)
+	{
+		const CSpell * sourceSpell = bonus->sourceSpell();
+		if(sourceSpell != nullptr)
+		{
+			//Special case: DISRUPTING_RAY is "immune" to dispell
+			//Other even PERMANENT effects can be removed (f.e. BIND)						
+			if(sourceSpell->id == SpellID::DISRUPTING_RAY)
```
Comment (DjWarmonger) [inline lib/spells/CDefaultSpellMechanics.cpp]: Okay now I see.

**Comment by ArseniyShestakov** [comment]
Now issue with GUI freeze fixed.

Played several maps on that branch and everything looks stable so far.

**Comment by ArseniyShestakov** [comment]
BTW there is no issues with archangels actually.
This was my fault because I wasn't able to reproduce it.
Likely I just used it before and forgot about it, etc.

**Pull request merged**